<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="NextGenMap: Fast and accurate read mapping in highly polymorphic genomes" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>NextGenMap</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/Cibiv/NextGenMap">View on GitHub</a>

          <h1 id="project_title">Nextgenmap</h1>
          <h2 id="project_tagline">A mapping method for Next Generation Sequencing reads</h2>

            <!--<section id="downloads">
              <a class="zip_download_link" href="http://github.com/Cibiv/NextGenMap/archive/v0.4.9.zip">Download this project as a .zip file</a>
              <a class="tar_download_link" href="http://github.com/Cibiv/NextGenMap/archive/v0.4.9.tar.gz">Download this project as a tar.gz file</a>
            </section>-->
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>Welcome to NextGenMap.</h3>

<p>NextGenMap (NGM) is a flexible and fast read mapping program that is more than twice as fast as BWA, while achieving 
a mapping sensitivity similar to Stampy or Bowtie2. NextGenMap uses a memory efficient index structure (hash table) to
store the positions of all 13-mers present in the reference genome. This index enables a quick identification of 
potential mapping regions for every read. Unlike other methods, NextGenMap dynamically determines for each read 
individually how many of the potential mapping regions have to be evaluated by a pairwise sequence alignment. 
Moreover, NextGenMap uses fast SIMD instructions (SSE) to accelerate the alignment calculations on the CPU. If
available NextGenMap calculates the alignments on the GPU (using OpenCL/CUDA) resulting in a runtime reduction 
of another 20 - 50 %, depending on the underlying data set.
<p></p>
Our results show that NextGenMap using only the CPU is at least twice as fast as BWA. Using the GPU for the alignment 
calculations increases the speedup to a factor of three. NextGenMap (GPU) even outperforms Bowtie2 by 10 - 50 % in 
terms of runtime. More importantly, the number of correctly mapped reads is similar to Stampy, one of the most 
sensitive methods available. 
</p>

<h3>Availability</h3>

<p>NextGenMap 0.5.0 (2014/10/17):
<p><a href="http://github.com/Cibiv/NextGenMap/archive/0.5.0.zip">Download this project as a .zip file</a></p>
<p><a href="http://github.com/Cibiv/NextGenMap/archive/0.5.0.tar.gz">Download this project as a tar.gz file</a></p>
<!--If you experience problems with the current release please try switching back to version 0.4.4 (<a href="https://github.com/Cibiv/NextGenMap/archive/0.4.4.tar.gz">tar</a>, <a href="https://github.com/Cibiv/NextGenMap/archive/0.4.4.zip">zip</a>)-->
</p>

<h3>Citation</h3>
<p>NextGenMap: fast and accurate read mapping in highly polymorphic genomes;
Fritz J. Sedlazeck, Philipp Rescheneder, Arndt von Haeseler;
Bioinformatics, Vol. 29, No. 21. (1 November 2013), pp. 2790-2791, doi:10.1093/bioinformatics/btt468 </p>
<a href="http://www.citeulike.org/user/stajich/article/12601670"> CiteULike entry</a> 


<h3>Changelog</h3>
Please see <a href="https://github.com/Cibiv/NextGenMap/wiki#changelog">here</a>

<h3>Used in publications</h3>
<p>1. Nicolas et. al. Finding and characterizing repeats in plant genomes (2016) <b>Plant Bioinformatics</b></p>
<p>2. Wu et. al. Bitpacking techniques for indexing genomes: II. Enhanced suffix arrays (2016) <b>Algorithms for Molecular Biology</b></p>
<p>3. Neme et. al. Fast turnover of genome transcription across evolutionary time exposes entire non-coding DNA to de novo gene emergence (2016) <b>eLife</b></p>
<p>4. Celorio-Mancera et. al. Evolutionary history of host use, rather than plant phylogeny, determines gene expression in a generalist butterfly (2016) <b>BMC Evolutionary Biology</b></p>
<p>5. Dubin et. al. DNA methylation in Arabidopsis has a genetic basis and shows evidence of local adaptation (2015) <b>eLife</b></p>
<p>6. Drecktrah et. al.  The Borrelia burgdorferi RelA/SpoT homolog and stringent response regulate survival in the tick vector and global gene expression during starvation (2015) <b>PLoS Pathog</b></p>
<p>7. Boisson-Dernier et. al. Receptor-like cytoplasmic kinase MARIS functions downstream of CrRLK1L-dependent signaling during tip growth (2015) <b>PNAS</b></p>
<p>8. Sun et. al. Transcription dynamically patterns the meiotic chromosome-axis interface <b>eLife</b></p>
<p>9. Huylmans et. al. Variation in the X: autosome distribution of male-biased genes among Drosophila melanogaster tissues and its relationship with dosage compensation (2015) <b>Genome biology and evolution</b></p>
<p>10. Sapeta et. al. Transcriptomics and physiological analyses reveal co-ordinated alteration of metabolic pathways in Jatropha curcas drought tolerance (2015) <b>J. Exp. Bot.</b></p>
<p>11. Tscherner et. al. The Candida albicans Histone Acetyltransferase Hat1 Regulates Stress Resistance and Virulence via Distinct Chromatin Assembly Pathways (2015) <b>PLoS Pathog</b></p>
<p>12. Muraya et. al. Targeted Sequencing Reveals Large-Scale Sequence Polymorphism in Maize Candidate Genes for Biomass Production and Composition (2015) <b>PLoS one</b></p>
<p>13. Paparazzo et. al. Survival Rate and Transcriptional Response upon Infection with the Generalist Parasite Beauveria bassiana in a World-Wide Sample of Drosophila melanogaster (2015) <b>PLoS one</b></p>
<p>14. Smolka et. al. Teaser: Individualized benchmarking and optimization of read mapping results for NGS data (2015) <b>Genome biology</b></p>
<p>15. Lindner et. al. Metagenomic profiling of known unknown microbes with MicrobeGPS (2015) <b>PLoS ONE</b></p>
<p>16. Krunic et. al.  Decreased expression of endogenous feline leukemia virus in cat lymphomas: a case control study (2015) <b>BMC Veterinary Research</b></p>
<p>17. Preusser et. al. Spectrum of gene mutations detected by next generation exome sequencing in brain metastases of lung adenocarcinoma (2015) <b>European Journal of Cancer</b></p>
<p>18. Ye et. al.  Alignment of Short Reads: A Crucial Step for Application of Next-Generation Sequencing Data in Precision Medicine (2015) <b>Pharmaceutics</b></p>
<p>19. Kang et al. Genome sequence of mungbean and insights into evolution within Vigna species (2014) <b>Nature Communications</b></p>
<p>20. Vesely et al. ADAR2 induces reproducible changes in sequence and abundance of mature microRNAs in the mouse brain (2014) <b>Nucleic Acids Research</b></p>
<p>21. Turki et. al. MaxSSmap: A GPU program for mapping divergent short reads to genomes with the maximum scoring subsequence(2014) <b>BMC Genomics</b></p>
<p>22. KÃ¶ster et. al. Massively parallel read mapping on GPUs with the q-group index and PEANUT(2014)<b> PeerJ</b></p>
<p>23. Tran et. al. Objective and comprehensive evaluation of bisulfite short read mapping tools(2014) <b>Advances in Bioinformatics</b></p> 
<p>24. Bilusic  et. al. Revisiting the coding potential of the E. coli genome through Hfq co-immunoprecipitation(2014) <b>RNA Biology</b></p>
<p>25. Huylmans  et. al. Population- and sex-biased gene expression in the excretion organs of Drosophila melanogaster(2014)<b> G3: Genes, Genomes, Genetics</b></p>
<p>26. Wright et. al. RAMICS: Trainable, high-speed and biologically relevant alignment of high-throughput sequencing reads to coding DNA(2014) <b>Nucleic Acids Research</b></p>
<p>27. McLaughlin et. al. Positive Selection and Multiple Losses of the LINE-1-Derived L1TD1 Gene in Mammals Suggest a Dual Role in Genome Defense and Pluripotency(2014) <b>PLoS Genetics</b></p>

<h3>Documentation</h3>

<p>You can find detailed information on how to install and use NextGenMap on our GitHub Wiki. Please follow this 
<a href="https://github.com/Cibiv/NextGenMap/wiki/">link to access the Wiki</a>.
</p>

<h3>Contact</h3>

<p>We are always intrested in feedback on how we can improve NextGenMap. If you have any suggestions please do not hesitate to 
either leave a note in the <a href="https://github.com/Cibiv/NextGenMap/issues/">issue section</a> or to contact us directly:</p>
<p></p><a href="mailto:fritz.sedlazeck@gmail.com">fritz.sedlazeck@gmail.com</a></p>
<p></p><a href="mailto:philipp.rescheneder@univie.ac.at">philipp.rescheneder@univie.ac.at</a></p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Nextgenmap maintained by <a href="https://github.com/Cibiv">Cibiv</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-49678031-1', 'cibiv.github.io');
  ga('send', 'pageview');

</script>


  </body>
</html>
