<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Nextgenmap : A mapping method for Next Generation Sequencing reads" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Nextgenmap</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/Cibiv/NextGenMap">View on GitHub</a>

          <h1 id="project_title">Nextgenmap</h1>
          <h2 id="project_tagline">A mapping method for Next Generation Sequencing reads</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/Cibiv/NextGenMap/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/Cibiv/NextGenMap/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>Welcome to NextGenMap.</h3>

<p>NextGenMap (NGM) is a flexible and fast read mapping program that is more than twice as fast as BWA, while achieving 
a mapping sensitivity similar to Stampy or Bowtie2. NextGenMap uses a memory efficient index structure (hash table) to
store the positions of all 13-mers present in the reference genome. This index enables a quick identification of 
potential mapping regions for every read. Unlike other methods, NextGenMap dynamically determines for each read 
individually how many of the potential mapping regions have to be evaluated by a pairwise sequence alignment. 
Moreover, NextGenMap uses fast SIMD instructions (SSE) to accelerate the alignment calculations on the CPU. If
available NextGenMap calculates the alignments on the GPU (using OpenCL/CUDA) resulting in a runtime reduction 
of another 20 - 50 %, depending on the underlying data set.
<p></p>
Our results show that NextGenMap using only the CPU is at least twice as fast as BWA. Using the GPU for the alignment 
calculations increases the speedup to a factor of three. NextGenMap (GPU) even outperforms Bowtie2 by 10 - 50 % in 
terms of runtime. More importantly, the number of correctly mapped reads is similar to Stampy, one of the most 
sensitive methods available. 
</p>

<h3>Availability</h3>

<p>NextGenMap 0.4.6:
<p></p></p><a href="https://github.com/Cibiv/NextGenMap/zipball/master">Download this project as a .zip file</a></p>
<p></p><a href="https://github.com/Cibiv/NextGenMap/tarball/master">Download this project as a tar.gz file</a></p>
If you experience problems with the current release please try switching back to version 0.4.4 (<a href"https://github.com/Cibiv/NextGenMap/archive/0.4.4.tar.gz">tar</a>, <a href"https://github.com/Cibiv/NextGenMap/archive/0.4.4.zip">zip</a>)
</p>

<h3>Changelog</h3>
<ul>
  <li>0.4.6: Set upper limit for max. number of reference sequences to 32,000 (will be increased in next version), switched to AMD APP SDK 2.8.1, fixed minor issue in ngm-utils</li>
  <li>0.4.5: Changed how NextGenMap handles multiple threads. This might cause problems. If you experience problems please switch back to 0.4.4!, changed default values for match, mismatch and gap penalties.</li>
  <li>0.4.4: Added dependencies (bamtools,...) to the source code, proper error handling for missing input files, mapping statistics (number of mapped, unmapped and written reads) fixed, print OpenCL driver version, fixed a bug that caused NGM to fail when mapping with theta close to 0, when computing theta the results are now rounded before converted to integer, NGM now handles reads that have no sequence, NGM can now handle references that consist of >32k sequences, "Invalid CIGAR string" and "Invalid reference position" errors where converted to warning, ngm-utils added</li>
  <li>0.4.3: Added BS-mapping cutoff (default 8), Fix for max-consec-indels and silent/hard clip</li>
  <li>0.4.2: Reduced memory usage when building hash table, improved parameter estimation, release/debug targets (CMAKE), paired-end flags for SAM/BAM output.</li>
  <li>0.4.1: Fixes a problem (segfault) when building the hash-table for large (> 2GB) genomes</li>
  <li>0.4.0: MASon and NextGenMap merged. CPU based bisulfite mapping added. Minor increase in mapping sensitivity Several minor bugfixes.</li>
  <li>0.3.1.1: Minor fix to work with gcc 4.5+.</li>
  <li>0.3.1: Improved bs-mapping.</li>
  <li>0.3.0: New release.</li>
  <li>0.2.0: Initial release.</li>
</ul>


<h3>Documentation</h3>

<p>You can find detailed information on how to install and use NextGenMap on our GitHub Wiki. Please follow this 
<a href="https://github.com/Cibiv/NextGenMap/wiki/">link to access the Wiki</a>.
</p>

<h3>Contact</h3>

<p>We are always intrested in feedback on how we can improve NextGenMap. If you have any suggestions please do not hesitate to 
either leave a note in the <a href="https://github.com/Cibiv/NextGenMap/issues/">issue section</a> or to contact us directly:</p>
<p></p><a href="mailto:fritz.sedlazeck@univie.ac.at">fritz.sedlazeck@univie.ac.at</a></p>
<p></p><a href="mailto:philipp.rescheneder@univie.ac.at">philipp.rescheneder@univie.ac.at</a></p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Nextgenmap maintained by <a href="https://github.com/Cibiv">Cibiv</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
