<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>BamAlignmentRecord</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Class_B_A_M+_I_2f_O.html#BamAlignmentRecord"></iframe></td><td valign=top><div id=content><div class=page_category>Class</div><div class=page_title>BamAlignmentRecord</div><div id=summary><div class=text>Represent a record from a BAM/SAM file.</div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/bam_io.h</div></div><div class=section id=remarks><div class=section_headline>Remarks</div><div class=text_block><div class=text_sub_block> While also used to represent SAM records, called <span class=code><nobr>BamAlignmentRecord</nobr></span> since the data directly reflects a BAM record (0-based positions, identify references by ids, not names, tags stored in BAM format.)</div></div></div><div class="section" id="memvar"><div class="section_headline">Data Members</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23begin_Pos.html" title="The position of this fragment mapping (0-based, INVALID_POS for '*')." >beginPos</a></nobr></td><td class="value_text" valign="top">The position of this fragment mapping (0-based, <span class=code><nobr>INVALID_POS</nobr></span> for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23bin.html" title="The bin of the alignment, automatically computed when writing BAM." >bin</a></nobr></td><td class="value_text" valign="top">The bin of the alignment, automatically computed when writing BAM.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23cigar.html" title="The CIGAR string as string of Class.CigarElement objects (empty for '*')." >cigar</a></nobr></td><td class="value_text" valign="top">The CIGAR string as string of <a href="CLASS_Cigar_Element.html" title="One entry of a CIGAR string." >CigarElement</a> objects (empty for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23flag.html" title="The flag of this mapping, see Enum.BamFlags for flag constants and the hasFlag* functions." >flag</a></nobr></td><td class="value_text" valign="top">The flag of this mapping, see <a href="ENUM_Bam_Flags.html" title="Shortcuts to the bitmask flags for BAM/SAM files." >BamFlags</a> for flag constants and the <span class=code><nobr>hasFlag*</nobr></span> functions.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23_I_N_V_A_L_I_D___L_E_N.html" title="Static member with invalid/sentinel position value." >INVALID_LEN</a></nobr></td><td class="value_text" valign="top">Static member with invalid/sentinel position value.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23_I_N_V_A_L_I_D___P_O_S.html" title="Static member with invalid/sentinel position value." >INVALID_POS</a></nobr></td><td class="value_text" valign="top">Static member with invalid/sentinel position value.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23_I_N_V_A_L_I_D___R_E_F_I_D.html" title="Static member with invalid/sentinel reference id (-1 as in BAM/SAM)." >INVALID_REFID</a></nobr></td><td class="value_text" valign="top">Static member with invalid/sentinel reference id (-1 as in BAM/SAM).</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23map_Q.html" title="The mapping quality (255 for '*')." >mapQ</a></nobr></td><td class="value_text" valign="top">The mapping quality (255 for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23p_Next.html" title="Position of next fragment mapping (0-based, INVALID_POS for '*')" >pNext</a></nobr></td><td class="value_text" valign="top">Position of next fragment mapping (0-based, <span class=code><nobr>INVALID_POS</nobr></span> for '*')</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23q_Name.html" title="The read/query name." >qName</a></nobr></td><td class="value_text" valign="top">The read/query name.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23qual.html" title="String with Phred scores (as in SAM file, empty for '*')." >qual</a></nobr></td><td class="value_text" valign="top">String with Phred scores (as in SAM file, empty for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23r_I_D.html" title="ID of reference for this fragment mapping (0-based, INVALID_REFID for '*')." >rID</a></nobr></td><td class="value_text" valign="top">ID of reference for this fragment mapping (0-based, <span class=code><nobr>INVALID_REFID</nobr></span> for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23r_Next_Id.html" title="ID of reference for next fragment mapping (0-based, INVALID_REFID for '*')" >rNextId</a></nobr></td><td class="value_text" valign="top">ID of reference for next fragment mapping (0-based, <span class=code><nobr>INVALID_REFID</nobr></span> for '*')</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23seq.html" title="The sequence string (empty for '*')." >seq</a></nobr></td><td class="value_text" valign="top">The sequence string (empty for '*').</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23tags.html" title="Raw BAM tag string, use Class.BamTagsDict for comfortable access." >tags</a></nobr></td><td class="value_text" valign="top">Raw BAM tag string, use <a href="CLASS_Bam_Tags_Dict.html" title="Indexes start positions of BAM tags in a Shortcut.CharString and provides a dict-like API." >BamTagsDict</a> for comfortable access.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Bam_Alignment_Record_23t_Len.html" title="The inferred template size (INVALID_LEN for '*')" >tLen</a></nobr></td><td class="value_text" valign="top">The inferred template size (<span class=code><nobr>INVALID_LEN</nobr></span> for '*')</td></tr></table></div><div class="section" id="memfunc"><div class="section_headline">Member Functions</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="MEMFUNC_Bam_Alignment_Record_23_Bam_Alignment_Record.html" title="Constructor." >BamAlignmentRecord</a></nobr></td><td class="value_text" valign="top">Constructor.</td></tr></table></div><div class="section" id="function"><div class="section_headline">Functions</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.bam_Record_To_Alignment.html" title="Convert Class.BamAlignmentRecord to an Class.Align object." >bamRecordToAlignment</a></nobr></td><td class="value_text" valign="top">Convert <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> to an <a href="CLASS_Align.html" title="An alignment of sequences." >Align</a> object.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.clear.html" title="Resets an object." >clear</a></nobr></td><td class="value_text" valign="top">Resets an object.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.get_Alignment_Length_In_Ref.html" title="Returns length of Class.BamAlignmentRecord's projection in reference." >getAlignmentLengthInRef</a></nobr></td><td class="value_text" valign="top">Returns length of <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a>'s projection in reference.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_All_Proper.html" title="Return true if a Class.BamAlignmentRecord has the &quot;all properly aligned&quot; flag set." >hasFlagAllProper</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "all properly aligned" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Duplicate.html" title="Return true if a Class.BamAlignmentRecord has the &quot;PCR or optical duplicate&quot; flag set." >hasFlagDuplicate</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "PCR or optical duplicate" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_First.html" title="Return true if a Class.BamAlignmentRecord has the &quot;first fragment of template&quot; flag set." >hasFlagFirst</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "first fragment of template" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Last.html" title="Return true if a Class.BamAlignmentRecord has the &quot;last fragment of template&quot; flag set." >hasFlagLast</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "last fragment of template" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Multiple.html" title="Return true if a Class.BamAlignmentRecord has the &quot;multiple&quot; flag set." >hasFlagMultiple</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "multiple" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Next_R_C.html" title="Return true if a Class.BamAlignmentRecord has the &quot;next fragment reverse-complemented&quot; flag set." >hasFlagNextRC</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "next fragment reverse-complemented" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Next_Unmapped.html" title="Return true if a Class.BamAlignmentRecord has the &quot;next fragment unmapped&quot; flag set." >hasFlagNextUnmapped</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "next fragment unmapped" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Q_C_No_Pass.html" title="Return true if a Class.BamAlignmentRecord has the &quot;does not pass quality controls&quot; flag set." >hasFlagQCNoPass</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "does not pass quality controls" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_R_C.html" title="Return true if a Class.BamAlignmentRecord has the &quot;reverse-complemented&quot; flag set." >hasFlagRC</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "reverse-complemented" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Secondary.html" title="Return true if a Class.BamAlignmentRecord has the &quot;secondary alignment&quot; flag set." >hasFlagSecondary</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "secondary alignment" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.has_Flag_Unmapped.html" title="Return true if a Class.BamAlignmentRecord has the &quot;fragment unmapped&quot; flag set." >hasFlagUnmapped</a></nobr></td><td class="value_text" valign="top">Return true if a <a href="CLASS_Bam_Alignment_Record.html" title="Represent a record from a BAM/SAM file." >BamAlignmentRecord</a> has the "fragment unmapped" flag set.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.read_Record.html" title="reads one record (e.g. a single DNA-sequence and its meta data) from a Concept.StreamConcept, by the means of Class.RecordReader" >readRecord</a></nobr></td><td class="value_text" valign="top">reads one record (e.g. a single DNA-sequence and its meta data) from a <a href="CONCEPT_Stream_Concept.html" title="Concept for I/O streams." >StreamConcept</a>, by the means of <a href="CLASS_Record_Reader.html" title="Buffer management for streams." >RecordReader</a></td></tr></table></div><div class="section" id="see"><div class="section_headline">See Also</div><div class="text_block"><span class="link_text"><a href="ENUM_Bam_Flags.html" title="Shortcuts to the bitmask flags for BAM/SAM files." >BamFlags</a></span></div></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:43:55</p></body></html>