<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>AlignedReadStoreElement</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Class_Fragment+_Store.html#AlignedReadStoreElement"></iframe></td><td valign=top><div id=content><div class=page_category>Class</div><div class=page_title>AlignedReadStoreElement</div><div id=summary><div class=text>Represents an alignment between read and contig.</div></div><div class=section id=signature><div class=signature_block><nobr><span class=code><nobr>AlignedReadStoreElement&lt;&gt;</nobr></span></nobr></div><div class=signature_block><nobr><span class=code><nobr>AlignedReadStoreElement&lt;TPos[,&nbsp;TGapAnchor[,&nbsp;TSpec]]&gt;</nobr></span></nobr></div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/store.h</div></div><div class=section id=param><div class=section_headline>Parameters</div><table class=value_tab cellspacing=0 cellpadding=0><tr><td class=value_key valign=top><nobr>TPos</nobr></td><td class=value_text valign=top><div class=text_sub_block> Type to store (gap-space) positions.</div></td></tr><tr><td class=value_key valign=top><nobr>TGapAnchor</nobr></td><td class=value_text valign=top><div class=text_sub_block> Type of a read gap anchor.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Gap_Anchor.html" title="Stores the position of an alignment character in sequence-space and in gap-space." >GapAnchor</a></div></td></tr><tr><td class=value_key valign=top><nobr>TSpec</nobr></td><td class=value_text valign=top><div class=text_sub_block> The specialization type.</div><div class=text_sub_block><span class=section_sub_headline>Default:</span> <span class=code><nobr>void</nobr></span></div></td></tr></table></div><div class=section id=remarks><div class=section_headline>Remarks</div><div class=text_block><div class=text_sub_block> Value type of the <a href="MEMVAR_Fragment_Store_23aligned_Read_Store.html" title="Class.String that stores &lt;alignId, readId, contigId, pairMatchId, beginPos, endPos, gaps>." >alignedReadStore</a> string.
In contrast to all other <a href="CLASS_Fragment_Store.html" title="Multi container to store contigs, reads, multiple read alignments and genome annotations." >FragmentStore</a> stores, the <a href="MEMVAR_Aligned_Read_Store_Element_23id.html" title="The alignment id refers to associated alignment information in Memvar.FragmentStore#alignQualityStore or Memvar.FragmentStore#alignedReadTagStore." >id</a> of an aligned read is explicitly stored
as a member to allow for random reordering of the <a href="MEMVAR_Fragment_Store_23aligned_Read_Store.html" title="Class.String that stores &lt;alignId, readId, contigId, pairMatchId, beginPos, endPos, gaps>." >alignedReadStore</a>, e.g. sorting by genomic position via <a href="FUNCTION.sort_Aligned_Reads.html" title="Stably sort aligned reads." >sortAlignedReads</a>.</div></div></div><div class="section" id="typedef"><div class="section_headline">Typedefs</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="TYPEDEF_Aligned_Read_Store_Element_23_T_Gap_Anchors.html" title="Type of the Memvar.AlignedReadStoreElement#gaps member." >TGapAnchors</a></nobr></td><td class="value_text" valign="top">Type of the <a href="MEMVAR_Aligned_Read_Store_Element_23gaps.html" title="String of read gap anchors. Can be used to create a Spec.AnchorGaps alignment row." >gaps</a> member.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="TYPEDEF_Aligned_Read_Store_Element_23_T_Id.html" title="Type of all stored ids." >TId</a></nobr></td><td class="value_text" valign="top">Type of all stored ids.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="TYPEDEF_Aligned_Read_Store_Element_23_T_Pos.html" title="Type of the Memvar.AlignedReadStoreElement#beginPos and Memvar.AlignedReadStoreElement#endPos members." >TPos</a></nobr></td><td class="value_text" valign="top">Type of the <a href="MEMVAR_Aligned_Read_Store_Element_23begin_Pos.html" title="Begin position of the alignment in gap-space." >beginPos</a> and <a href="MEMVAR_Aligned_Read_Store_Element_23end_Pos.html" title="End position of the alignment in gap-space. If Memvar.AlignedReadStoreElement#endPos &lt; Memvar.AlignedReadStoreElement#beginPos, the read is aligned to the reverse strand, where Memvar.AlignedReadStoreElement#beginPos and Memvar.AlignedReadStoreElement#endPos are the corresponding positions on the forward strand." >endPos</a> members.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="TYPEDEF_Aligned_Read_Store_Element_23_T_Spec.html" title="The specialization type." >TSpec</a></nobr></td><td class="value_text" valign="top">The specialization type.</td></tr></table></div><div class="section" id="memvar"><div class="section_headline">Data Members</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23begin_Pos.html" title="Begin position of the alignment in gap-space." >beginPos</a></nobr></td><td class="value_text" valign="top">Begin position of the alignment in gap-space.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23contig_Id.html" title="Refers to the contig in the Memvar.FragmentStore#contigStore the read is aligned with." >contigId</a></nobr></td><td class="value_text" valign="top">Refers to the contig in the <a href="MEMVAR_Fragment_Store_23contig_Store.html" title="Class.String that maps from contigId to &lt;contigSeq, contigGaps, contigFileId>." >contigStore</a> the read is aligned with.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23end_Pos.html" title="End position of the alignment in gap-space. If Memvar.AlignedReadStoreElement#endPos &lt; Memvar.AlignedReadStoreElement#beginPos, the read is aligned to the reverse strand, where Memvar.AlignedReadStoreElement#beginPos and Memvar.AlignedReadStoreElement#endPos are the corresponding positions on the forward strand." >endPos</a></nobr></td><td class="value_text" valign="top">End position of the alignment in gap-space. If <a href="MEMVAR_Aligned_Read_Store_Element_23end_Pos.html" title="End position of the alignment in gap-space. If Memvar.AlignedReadStoreElement#endPos &lt; Memvar.AlignedReadStoreElement#beginPos, the read is aligned to the reverse strand, where Memvar.AlignedReadStoreElement#beginPos and Memvar.AlignedReadStoreElement#endPos are the corresponding positions on the forward strand." >endPos</a> &lt; <a href="MEMVAR_Aligned_Read_Store_Element_23begin_Pos.html" title="Begin position of the alignment in gap-space." >beginPos</a>, the read is aligned to the reverse strand,
where <a href="MEMVAR_Aligned_Read_Store_Element_23begin_Pos.html" title="Begin position of the alignment in gap-space." >beginPos</a> and <a href="MEMVAR_Aligned_Read_Store_Element_23end_Pos.html" title="End position of the alignment in gap-space. If Memvar.AlignedReadStoreElement#endPos &lt; Memvar.AlignedReadStoreElement#beginPos, the read is aligned to the reverse strand, where Memvar.AlignedReadStoreElement#beginPos and Memvar.AlignedReadStoreElement#endPos are the corresponding positions on the forward strand." >endPos</a> are the corresponding positions on the forward strand.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23gaps.html" title="String of read gap anchors. Can be used to create a Spec.AnchorGaps alignment row." >gaps</a></nobr></td><td class="value_text" valign="top">String of read gap anchors. Can be used to create a <a href="SPEC_Anchor_Gaps.html" title="Stores gaps anchors of the first characters behind gaps." >AnchorGaps</a> alignment row.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23id.html" title="The alignment id refers to associated alignment information in Memvar.FragmentStore#alignQualityStore or Memvar.FragmentStore#alignedReadTagStore." >id</a></nobr></td><td class="value_text" valign="top">The alignment id refers to associated alignment information in <a href="MEMVAR_Fragment_Store_23align_Quality_Store.html" title="Class.String that maps from alignId to &lt;pairScore, score, errors>." >alignQualityStore</a> or <a href="MEMVAR_Fragment_Store_23aligned_Read_Tag_Store.html" title="Class.StringSet that maps from alignId to alignTag." >alignedReadTagStore</a>.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23_I_N_V_A_L_I_D___I_D.html" title="Constant to represent an invalid id." >INVALID_ID</a></nobr></td><td class="value_text" valign="top">Constant to represent an invalid id.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23pair_Match_Id.html" title="Two read alignments having the same Memvar.AlignedReadStoreElement#pairMatchId form a valid pair match. If it equals Memvar.AlignedReadStoreElement#INVALID_ID, the read is either not paired or could not be aligned as part of a pair match." >pairMatchId</a></nobr></td><td class="value_text" valign="top">Two read alignments having the same <a href="MEMVAR_Aligned_Read_Store_Element_23pair_Match_Id.html" title="Two read alignments having the same Memvar.AlignedReadStoreElement#pairMatchId form a valid pair match. If it equals Memvar.AlignedReadStoreElement#INVALID_ID, the read is either not paired or could not be aligned as part of a pair match." >pairMatchId</a> form a valid pair match.
If it equals <a href="MEMVAR_Aligned_Read_Store_Element_23_I_N_V_A_L_I_D___I_D.html" title="Constant to represent an invalid id." >INVALID_ID</a>, the read is either not paired or could not be aligned as part of a pair match.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="MEMVAR_Aligned_Read_Store_Element_23read_Id.html" title="Refers to the aligned read in the Memvar.FragmentStore#readStore." >readId</a></nobr></td><td class="value_text" valign="top">Refers to the aligned read in the <a href="MEMVAR_Fragment_Store_23read_Store.html" title="Class.String that maps from readId to &lt;matePairId>." >readStore</a>.</td></tr></table></div><div class="section" id="memfunc"><div class="section_headline">Member Functions</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="MEMFUNC_Aligned_Read_Store_Element_23_Aligned_Read_Store_Element.html" title="Constructor" >AlignedReadStoreElement</a></nobr></td><td class="value_text" valign="top">Constructor</td></tr></table></div><div class="section" id="function"><div class="section_headline">Functions</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.get_Clr_Range.html" title="Get the &quot;clear&quot; range of a read alignment." >getClrRange</a></nobr></td><td class="value_text" valign="top">Get the "clear" range of a read alignment.</td></tr></table></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:43:55</p></body></html>