<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>Predefined Annotation Types</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Enum_Fragment+_Store.html#Predefined Annotation Types"></iframe></td><td valign=top><div id=content><div class=page_category>Enum</div><div class=page_title>Predefined Annotation Types</div><div id=summary><div class=text>Predefined annotation type ids.</div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/store.h</div></div><div class=section id=remarks><div class=section_headline>Remarks</div><div class=text_block><div class=text_sub_block> The <a href="CLASS_Fragment_Store.html" title="Multi container to store contigs, reads, multiple read alignments and genome annotations." >FragmentStore</a> predefines some commonly used <a href="MEMVAR_Annotation_Store_Element_23type_Id.html" title="Refers to an entry in the Memvar.FragmentStore#annotationTypeStore. There are some type ids predefined for commonly used types, e.g. ANNO_GENE. See Enum.Predefined Annotation Types." >typeId</a> values.
They can be used to compare or set the <a href="MEMVAR_Annotation_Store_Element_23type_Id.html" title="Refers to an entry in the Memvar.FragmentStore#annotationTypeStore. There are some type ids predefined for commonly used types, e.g. ANNO_GENE. See Enum.Predefined Annotation Types." >typeId</a> directly as a fast alternative to <a href="FUNCTION.get_Type.html" title="Returns the type name of the current annotation." >getType</a> and <a href="FUNCTION.set_Type.html" title="Sets the type name of the current annotation." >setType</a>.</div></div></div><div class="section" id="class"><div class="section_headline">Member of</div><div class="text_block"><span class="link_text"><a href="CLASS_Fragment_Store.html" title="Multi container to store contigs, reads, multiple read alignments and genome annotations." >FragmentStore</a></span></div></div><div class=section id=value><div class=section_headline>Values</div><table class=value_tab cellspacing=0 cellpadding=0><tr><td class=value_key valign=top><nobr>ANNO_ROOT</nobr></td><td class=value_text valign=top><div class=text_sub_block> The root node ("&lt;root&gt;").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_GENE</nobr></td><td class=value_text valign=top><div class=text_sub_block> A gene ("gene").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_MRNA</nobr></td><td class=value_text valign=top><div class=text_sub_block> An mRNA sequence, aka transcript ("mRNA").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_CDS</nobr></td><td class=value_text valign=top><div class=text_sub_block> A coding region ("CDS").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_EXON</nobr></td><td class=value_text valign=top><div class=text_sub_block> An exon ("exon").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_FIVE_PRIME_UTR</nobr></td><td class=value_text valign=top><div class=text_sub_block> A 5' untranslated region ("five_prime_UTR").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_INTRON</nobr></td><td class=value_text valign=top><div class=text_sub_block> An intron ("intron").</div></td></tr><tr><td class=value_key valign=top><nobr>ANNO_THREE_PRIME_UTR</nobr></td><td class=value_text valign=top><div class=text_sub_block> A 3' untranslated region ("three_prime_UTR").</div></td></tr></table></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:43:56</p></body></html>