<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>splitAlignment</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Function.html#splitAlignment"></iframe></td><td valign=top><div id=content><div class=page_category>Function</div><div class=page_title>splitAlignment</div><div id=summary><div class=text>Compute split alignments.</div></div><div class=section id=description><div class=text_block><div class=text_sub_block> 
There are two variants of the split alignment problem.
In the first variant, we wan to align two sequences where the first (say the reference) one is shorter than the second (say a read) and the read contains an insertion with respect to the reference.
We now want to align the read agains the reference such that the left part of the read aligns well against the left part of the reference and the right part of the read aligns well against the right part of the reference.
The center gap in the reference is free. For example:</div><div class=code_sub_block><span class=code><nobr>reference&nbsp;&nbsp;AGCATGTTAGATAAGATAGC-----------TGTGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;||||||||||||||||||||&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|||||||||||||||||||||||||<br >read&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGCATGTTAGATAAGATAGCCCCCCCCCCCCTGTGCTAGTAGGCAGTCAGCGCCAT</nobr></span></div><div class=text_sub_block> 
The second variant is to align two sequences A and B against a reference such that the left part of A aligns well to the left part of the reference and the right part of B aligns well to the right part of the reference.
Together, both reads span the whole reference and overlap with an insertion in the reference.</div><div class=code_sub_block><span class=code><nobr>reference&nbsp;&nbsp;AGCATGTTAGATAAGATAGCTGTGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;||||||||||||||||||&nbsp;|&nbsp;||<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGCATGTTAGATAAGATATCCGTCC<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read&nbsp;1<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|||&nbsp;|||||||||||||||||||||||<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CCGCTATGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read&nbsp;2</nobr></span></div><div class=text_sub_block> 
The resulting alignment of the left/right parts is depicted below.
The square brackets indicate clipping positions.</div><div class=code_sub_block><span class=code><nobr>reference&nbsp;&nbsp;AGCATGTTAGATAAGATA&nbsp;&nbsp;&nbsp;&nbsp;[GCTGTGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;||||||||||||||||||&nbsp;&nbsp;&nbsp;&nbsp;[&nbsp;|&nbsp;||<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;AGCATGTTAGATAAGATA&nbsp;&nbsp;&nbsp;&nbsp;[TCCGTCC<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read&nbsp;1<br >reference&nbsp;&nbsp;AGCATGTTAGATAAGATA]&nbsp;&nbsp;&nbsp;&nbsp;GTGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;]&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|||||||||||||||||||||||<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CCGCT]&nbsp;&nbsp;&nbsp;&nbsp;ATGCTAGTAGGCAGTCAGCGCCAT<br >&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read&nbsp;2</nobr></span></div><div class=text_sub_block> 
In the first case, we want to find the one breakpoint in the reference and the two breakpoints in the reads and the alignment of the left and right well-aligning read parts.
In the second case, we want to find the one breakpoint in the reference and the breakpoint/clipping position in each read.</div><div class=text_sub_block> 
The <span class=code><nobr>splitAlignment()</nobr></span> function takes as the input two alignments.
The sequence in each alignment's first row is the reference and the sequence of the second row is the read.
The sequence has to be the same sequence whereas the reads might differ.
If the reads are the same then this is the same as the first case and if the reads differ then this is the second case.</div><div class=text_sub_block> 
The result is two alignments of the left and right contig path clipped appropriately.
The resulting score is the sum of the scores of both alignments.</div></div></div><div class=section id=signature><div class=signature_block><nobr><span class=code><nobr>TScoreValue&nbsp;splitAlignment(alignL,&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;alignR,&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;scoringScheme[,&nbsp;lowerDiag,&nbsp;upperDiag])</nobr></span></nobr></div><div class=signature_block><nobr><span class=code><nobr>TScoreValue&nbsp;splitAlignment(gapsHL,&nbsp;gapsVL,&nbsp;gapsHR,&nbsp;gapsVR,&nbsp;scoringScheme[,&nbsp;lowerDiag,&nbsp;upperDiag])</nobr></span></nobr></div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/align_split.h</div></div><div class=section id=param><div class=section_headline>Parameters</div><table class=value_tab cellspacing=0 cellpadding=0><tr><td class=value_key valign=top><nobr>alignL</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Align.html" title="An alignment of sequences." >Align</a> object with two rows for the left alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Align.html" title="An alignment of sequences." >Align</a></div></td></tr><tr><td class=value_key valign=top><nobr>alignR</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Align.html" title="An alignment of sequences." >Align</a> object with two rows for the right alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Align.html" title="An alignment of sequences." >Align</a></div></td></tr><tr><td class=value_key valign=top><nobr>gapsHL</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a> object with the horizontal/contig row for the left alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a></div></td></tr><tr><td class=value_key valign=top><nobr>gapsVL</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a> object with the vertical/read row for the left alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a></div></td></tr><tr><td class=value_key valign=top><nobr>gapsHR</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a> object with the horizontal/contig row for the right alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a></div></td></tr><tr><td class=value_key valign=top><nobr>gapsVR</nobr></td><td class=value_text valign=top><div class=text_sub_block> <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a> object with the vertical/read row for the right alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Gaps.html" title="Efficient storage of gaps for a sequence." >Gaps</a></div></td></tr><tr><td class=value_key valign=top><nobr>scoringScheme</nobr></td><td class=value_text valign=top><div class=text_sub_block> The scoring scheme to use for the alignment.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Score.html" title="A scoring scheme." >Score</a></div></td></tr><tr><td class=value_key valign=top><nobr>lowerDiag</nobr></td><td class=value_text valign=top><div class=text_sub_block> The lower diagonal.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <span class=code><nobr>int</nobr></span></div><div class=text_sub_block><span class=section_sub_headline>Remarks:</span> You have to specify the upper and lower diagonals for the left alignment.  For the right alignment, the corresponding diagonals are chosen for the lower right part of the DP matrix.</div></td></tr><tr><td class=value_key valign=top><nobr>upperDiag</nobr></td><td class=value_text valign=top><div class=text_sub_block> The lower diagonal. Also see remark for <span class=code><nobr>lowerDiag</nobr></span>.</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <span class=code><nobr>int</nobr></span></div></td></tr></table></div><div class=section id=remarks><div class=section_headline>Remarks</div><div class=text_block><div class=text_sub_block> The DP algorithm is chosen automatically depending on whether the gap open and extension costs are equal.</div></div></div><div class=section id=returns><div class=section_headline>Return Values</div><div class=text_block><div class=text_sub_block> 
The sum of the alignment scores of both alignments.
<span class=code><nobr>TScoreValue</nobr></span> is the value type of <span class=code><nobr>scoringScheme</nobr></span>.</div></div></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:43:59</p></body></html>