<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>countKmers</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Function_Alignment+_Free.html#countKmers"></iframe></td><td valign=top><div id=content><div class=page_category>Function</div><div class=page_title>countKmers</div><div id=summary><div class=text>Counts kmers in a sequence. Optionally, a background model is returned.</div></div><div class=section id=signature><div class=signature_block><nobr><span class=code><nobr>countKmers(kmerCounts,&nbsp;sequence,&nbsp;k)</nobr></span></nobr></div><div class=signature_block><nobr><span class=code><nobr>countKmers(kmerCounts,&nbsp;backgroundFrequencies,&nbsp;sequence,&nbsp;k)</nobr></span></nobr></div><div class=signature_block><nobr><span class=code><nobr>countKmers(kmerCounts,&nbsp;bgModel,&nbsp;sequence,&nbsp;k)</nobr></span></nobr></div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/alignment_free.h</div></div><div class=section id=param><div class=section_headline>Parameters</div><table class=value_tab cellspacing=0 cellpadding=0><tr><td class=value_key valign=top><nobr>kmerCounts</nobr></td><td class=value_text valign=top><div class=text_sub_block> String&lt;unsigned&gt; with kmer counts for every k-mer</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_String.html" title="A sequence container with generic alphabet and many specializations." >String</a></div></td></tr><tr><td class=value_key valign=top><nobr>backgroundFrequencies</nobr></td><td class=value_text valign=top><div class=text_sub_block> String of background frequencies representing the model</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  double</div></td></tr><tr><td class=value_key valign=top><nobr>bgModel</nobr></td><td class=value_text valign=top><div class=text_sub_block> Markov model</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_Markov_Model.html" title="Gives a suitable representation of a Marcov Chain." >MarkovModel</a></div></td></tr><tr><td class=value_key valign=top><nobr>sequence</nobr></td><td class=value_text valign=top><div class=text_sub_block> String (sequence) where k-mers are counted</div><div class=text_sub_block><span class=section_sub_headline>Types:</span>  <a href="CLASS_String.html" title="A sequence container with generic alphabet and many specializations." >String</a></div></td></tr></table></div><div class=section id=remarks><div class=section_headline>Remarks</div><div class=text_block><div class=text_sub_block> k-mers overlapping masked letters are not counted in case of Dna5Strings. A Bernoulli or Markov Model can be choosen as a background model.</div></div></div><div class=section id=returns><div class=section_headline>Return Values</div><div class=text_block><div class=text_sub_block> String&lt;unsigned&gt; with kmer counts for every k-mer</div></div></div><div class=section id=example><div class=section_headline>Examples</div><div class=text_block><div class=text_sub_block>  Calculate the alignment free sequence similarity o two masked DNA sequences.</div><div class=code_sub_block><span class=code><nobr>using&nbsp;namespace&nbsp;seqan;<br >//&nbsp;Masked&nbsp;sequence,&nbsp;we&nbsp;do&nbsp;not&nbsp;want&nbsp;to&nbsp;count&nbsp;words&nbsp;overlapping&nbsp;'N'<br >Dna5String&nbsp;sequenceDna5&nbsp;=<br >&nbsp;&nbsp;&nbsp;&nbsp;"TAGGTTTTCCGAAAAGGTAGCAACTTTACGTGATCAAACCTCTGACGGGGTTTTCCCCGTCGAAATTGGGTG"<br >&nbsp;&nbsp;&nbsp;&nbsp;"TTTCTTGTCTTGTTCTCACTTGGGGCATCTCCGTCAAGCCAAGAAAGTGCTCCCTGGATTCTGTTGCTAACG"<br >&nbsp;&nbsp;&nbsp;&nbsp;"AGTCTCCTCTGCATTCCTGCTTGACTGATTGGGCGGACGGGGTGTCCACCTGACGCTGAGTATCGCCGTCAC"<br >&nbsp;&nbsp;&nbsp;&nbsp;"GGTGCCACATGTCTTATCTATTCAGGGATCAGAATTTATTCAGGAAATCAGGAGATGCTACACTTGGGTTAT"<br >&nbsp;&nbsp;&nbsp;&nbsp;"CGAAGCTCCTTCCAAGGCGTAGCAAGGGCGACTGAGCGCGTAAGCTCTAGATCTCCTCGTGTTGCAACTACA"<br >&nbsp;&nbsp;&nbsp;&nbsp;"CGCGCGGGTCACTCGAAACACATAGTATGAACTTAACGACTGCTCGTACTGAACAATGCTGAGGCAGAAGAT"<br >&nbsp;&nbsp;&nbsp;&nbsp;"CGCAGACCAGGCATCCCACTGCTTGAAAAAACTATNNNNCTACCCGCCTTTTTATTATCTCATCAGATCAAG";<br >&nbsp;<br >String&lt;unsigned&gt;&nbsp;kmerCounts;<br >unsigned&nbsp;k&nbsp;=&nbsp;2;&nbsp;&nbsp;//&nbsp;Count&nbsp;all&nbsp;2-mers<br >countKmers(kmerCounts,&nbsp;sequenceDna5,&nbsp;k);<br >&nbsp;<br >for(unsigned&nbsp;i&nbsp;=&nbsp;0;&nbsp;i&lt;16;&nbsp;++i)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//&nbsp;Print&nbsp;the&nbsp;2-mer&nbsp;counts<br >&nbsp;&nbsp;&nbsp;&nbsp;std::cout&lt;&lt;kmerCounts[i]&lt;&lt;"\n";&nbsp;&nbsp;//&nbsp;34&nbsp;times&nbsp;AA;&nbsp;30&nbsp;times&nbsp;AC;&nbsp;28&nbsp;times&nbsp;AG;&nbsp;...<br >&nbsp;<br >&nbsp;<br >String&lt;double&gt;&nbsp;nucleotideFrequencies;&nbsp;&nbsp;//&nbsp;Defines&nbsp;a&nbsp;Bernoulli&nbsp;model&nbsp;for&nbsp;DNA&nbsp;sequences.<br >//&nbsp;Count&nbsp;all&nbsp;2-mers&nbsp;and&nbsp;save&nbsp;the&nbsp;nucleotide&nbsp;frequencies<br >countKmers(kmerCounts,&nbsp;nucleotideFrequencies,&nbsp;sequenceDna5,&nbsp;k);<br >&nbsp;<br >for(unsigned&nbsp;i&nbsp;=&nbsp;0;&nbsp;i&lt;4;&nbsp;++i)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//&nbsp;Print&nbsp;the&nbsp;nucleotide&nbsp;frequencies<br >&nbsp;&nbsp;&nbsp;&nbsp;std::cout&nbsp;&lt;&lt;&nbsp;nucleotideFrequencies[i]&nbsp;&lt;&lt;&nbsp;"\n";<br >//&nbsp;=&gt;&nbsp;p(A)&nbsp;=&nbsp;0.238;&nbsp;p(C)&nbsp;=&nbsp;0.254;&nbsp;p(G)&nbsp;=&nbsp;0.238;&nbsp;p(T)&nbsp;=&nbsp;0.27;<br >&nbsp;<br >MarkovModel&lt;Dna,&nbsp;double&gt;&nbsp;&nbsp;backgroundModel(1);&nbsp;&nbsp;//&nbsp;Markov&nbsp;model&nbsp;of&nbsp;order&nbsp;1<br >//&nbsp;Count&nbsp;all&nbsp;2-mers&nbsp;and&nbsp;return&nbsp;a&nbsp;Markov&nbsp;model<br >countKmers(kmerCounts,&nbsp;backgroundModel,&nbsp;sequenceDna5,&nbsp;k);<br >std::cout&lt;&lt;backgroundModel.transition;&nbsp;&nbsp;//&nbsp;Print&nbsp;the&nbsp;transition&nbsp;matrix&nbsp;of&nbsp;the&nbsp;markov&nbsp;model</nobr></span></div></div></div><div class="section" id="see"><div class="section_headline">See Also</div><div class="text_block"><span class="link_text"><a href="CLASS_Markov_Model.html" title="Gives a suitable representation of a Marcov Chain." >MarkovModel</a></span>, <span class="link_text"><a href="FUNCTION.alignment_Free_Comparison.html" title="Computes the pairwise similarity scores for a set of sequences" >alignmentFreeComparison</a></span>, <span class="link_text"><a href="FUNCTION.string_To_String_Set.html" title="Transform a String into a StringSet containing this String." >stringToStringSet</a></span>, <span class="link_text"><a href="FUNCTION.calculate_Probability.html" title="Calculates the probability of a sequence given a Bernoulli model (String of background frequencies)" >calculateProbability</a></span>, <span class="link_text"><a href="FUNCTION.calculate_Covariance.html" title="Calculates the covariance for the number of word occurrences for two words in a sequence of length n, given a background model." >calculateCovariance</a></span>, <span class="link_text"><a href="FUNCTION.calculate_Variance.html" title="Calculates the variance for the number of word occurrences of a word in a sequence of length n given a background model." >calculateVariance</a></span></div></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:43:57</p></body></html>