<html><head><meta http-equiv="content-type" content="text/html; charset=UTF-8"><link rel="stylesheet" href="dddoc_html.css" type="text/css" /><title>File Format</title></head><body><table id=main_table cellspacing=0 cellpadding=0><tr><td valign=top><iframe frameborder=0 id=navigation src="INDEX_Tag_Input_2f_Output.html#File Format"></iframe></td><td valign=top><div id=content><div class=page_category>Tag</div><div class=page_title>File Format</div><div id=summary><div class=text>A file format.</div></div><div class=section id=include><div class=section_headline>Include Headers</div><div class=text_block>seqan/file.h</div></div><div class="section" id="function"><div class="section_headline">Functions</div><table class="value_tab" cellspacing="0" cellpadding="0"><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.append_Seqs.html" title="Appends all sequences stored in files of directory to a StringSet." >appendSeqs</a></nobr></td><td class="value_text" valign="top">Appends all sequences stored in files of directory to a StringSet.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.assign_Cropped_Seq_Id.html" title="Extracts the sequence id up to the first whitespace of a sequence file fragment." >assignCroppedSeqId</a></nobr></td><td class="value_text" valign="top">Extracts the sequence id up to the first whitespace of a sequence file fragment.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.assign_Qual.html" title="Extracts the quality values of a sequence file fragment." >assignQual</a></nobr></td><td class="value_text" valign="top">Extracts the quality values of a sequence file fragment.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.assign_Qual_Id.html" title="Extracts the quality value id of a sequence file fragment." >assignQualId</a></nobr></td><td class="value_text" valign="top">Extracts the quality value id of a sequence file fragment.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.assign_Seq.html" title="Extracts the sequence part of a sequence file fragment." >assignSeq</a></nobr></td><td class="value_text" valign="top">Extracts the sequence part of a sequence file fragment.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.assign_Seq_Id.html" title="Extracts the sequence id of a sequence file fragment." >assignSeqId</a></nobr></td><td class="value_text" valign="top">Extracts the sequence id of a sequence file fragment.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.guess_Format.html" title="Guesses a file format from the contents of a sequence file." >guessFormat</a></nobr></td><td class="value_text" valign="top">Guesses a file format from the contents of a sequence file.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.guess_Format_From_Filename.html" title="Guesses a file format from a sequence file name." >guessFormatFromFilename</a></nobr></td><td class="value_text" valign="top">Guesses a file format from a sequence file name.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.guess_Stream_Format.html" title="check whether the data provided by reader is (one of) the specified format(s)." >guessStreamFormat</a></nobr></td><td class="value_text" valign="top">check whether the data provided by reader is (one of) the specified format(s).</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.print_Alignment.html" title="Prints a window of the visible layout of reads into a outstream." >printAlignment</a></nobr></td><td class="value_text" valign="top">Prints a window of the visible layout of reads into a outstream.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.read.html" title="Loads records from a file." >read</a></nobr></td><td class="value_text" valign="top">Loads records from a file.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.split.html" title="Divides the contents of a sequence file into sequence file fragments separated by a file format specific delimiter." >split</a></nobr></td><td class="value_text" valign="top">Divides the contents of a sequence file into sequence file fragments separated by a file format specific delimiter.</td></tr><tr><td class="value_key" valign="top"><nobr><a href="FUNCTION.write.html" title="Saves records to a file." >write</a></nobr></td><td class="value_text" valign="top">Saves records to a file.</td></tr></table></div><div class=section id=tag><div class=section_headline>Tags</div><table class=value_tab cellspacing=0 cellpadding=0><tr><td class=value_key valign=top><nobr>CGViz</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	CGViz file format for sequences. Only output.</div></td></tr><tr><td class=value_key valign=top><nobr>Embl</nobr></td><td class=value_text valign=top><div class=text_sub_block> EMBL format for sequences.</div></td></tr><tr><td class=value_key valign=top><nobr>Fasta</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	FASTA file format for sequences.</div><div class=text_sub_block> 
    FASTA file format for sequences.</div></td></tr><tr><td class=value_key valign=top><nobr>Fasta alignment</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	FASTA alignment file format for sequences.</div></td></tr><tr><td class=value_key valign=top><nobr>Genbank</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	Genbank format for sequences from the Genbank database.</div></td></tr><tr><td class=value_key valign=top><nobr>Fastq</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
    FASTQ file format for sequences.</div><div class=text_sub_block> 
    FASTQ file format for sequences.</div></td></tr><tr><td class=value_key valign=top><nobr>QSeq</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	QSeq format, used for most of the Illumina read files.</div></td></tr><tr><td class=value_key valign=top><nobr>Raw</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	Raw file format for sequences.</div><div class=text_sub_block> 
	The file contains data in a raw format.</div></td></tr><tr><td class=value_key valign=top><nobr>Gff</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
    Gff annotation file.</div></td></tr><tr><td class=value_key valign=top><nobr>Gtf</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
    Gtf annotation file.</div></td></tr><tr><td class=value_key valign=top><nobr>ScoreMatrixFile</nobr></td><td class=value_text valign=top><div class=text_sub_block> Score matrix file.</div></td></tr><tr><td class=value_key valign=top><nobr>Amos message file</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
	Amos message file.</div></td></tr><tr><td class=value_key valign=top><nobr>Ucsc</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
    Ucsc Genome Browser annotation file (a.k.a. knownGene format).</div></td></tr><tr><td class=value_key valign=top><nobr>UcscIsoforms</nobr></td><td class=value_text valign=top><div class=text_sub_block> 
    Ucsc Genome Browser isoforms file (a.k.a. knownIsoforms format).</div></td></tr></table></div><div class="section" id="see"><div class="section_headline">See Also</div><div class="text_block"><span class="link_text"><a href="FUNCTION.guess_File_Format.html" title="Tries to determine the format of a file." >guessFileFormat</a></span>, <span class="link_text"><a href="CLASS_File_Format.html" title="Object that stores a file format." >FileFormat</a></span></div></div><div class=page_footline>SeqAn - Sequence Analysis Library - <a href="http://www.seqan.de">www.seqan.de</a></div><div id=page_widthblock>&nbsp;</div></div></td></tr></table><p style="font-size:50%; color: #909090">Page built @2013/07/11 10:44:02</p></body></html>